pkg/updater/updater.go

```diff
@@ -432,3 +353,71 @@ func checkAvailableDiskSpace(fsDisk disk, paths ...string) error {
 	}
 	return nil
 }
+
+type requestKey int
+
+var requestStateKey requestKey
+
+// requestState represents the state of a task.
+type requestState struct {
+	Package string
+	ID      string
+	State   pbgo.TaskState
+	Err     *updaterErrors.UpdaterError
+}
+
+func newRequestContext(request remoteAPIRequest) context.Context {
+	return context.WithValue(context.Background(), requestStateKey, &requestState{
+		Package: request.Package,
+		ID:      request.ID,
+		State:   pbgo.TaskState_RUNNING,
+	})
+}
+
+func setRequestInvalid(ctx context.Context) {
+	state := ctx.Value(requestStateKey).(*requestState)
+	state.State = pbgo.TaskState_INVALID_STATE
+}
```

Comment:
<!-- Warning represent user-facing problems. If possible, it's great to follow this kind of comment with a suggestion.  -->
**🔨 Warning:** ‏This doesn't feel safe to me:
1. If the context value is of the wrong type, we panic
2. If the context value ie somehow nil, we'll also panic

Reply:
I can add a check on the cast and do nothing if it fails. I don't know if it's better tho, because this would not be a regular error but an unexpected state/bug. In this context it might be dangerous to continue execution tho (we might report unexpected state to the backend or something worse)

Reply:
Panicking the updater shouldn't happen at all. Agreed we should report that somehow (telemetry?) but we could try to simply empty the context and reset all packages to IDLE.

Reply:
Discussed offline - given that we can't really agree yet on what to do if this scenario happens, and given that the updater is currently fully sequential and stateless (so it can recover), let's keep a panic in case of unexpected deref on the internal state.